src/components/mc/cuda/kernel/mc_cuda_reduce.cu

```diff
@@ -0,0 +1,168 @@
+        case UCC_OP_MAX:                                                       \
+            LAUNCH_KERNEL(MAX, type, sbuf1, sbuf2, dest, count, s, b, t);      \
+            break;                                                             \
+        case UCC_OP_MIN:                                                       \
+            LAUNCH_KERNEL(MIN, type, sbuf1, sbuf2, dest, count, s, b, t);      \
+            break;                                                             \
+        case UCC_OP_SUM:                                                       \
+            LAUNCH_KERNEL(SUM, type, sbuf1, sbuf2, dest, count, s, b, t);      \
+            break;                                                             \
+        case UCC_OP_PROD:                                                      \
+            LAUNCH_KERNEL(PROD, type, sbuf1, sbuf2, dest, count, s, b, t);     \
+            break;                                                             \
+        default:                                                               \
+            mc_error(&ucc_mc_cuda.super, "float dtype does not support "       \
+                                         "requested reduce op: %d", op);       \
+            return UCC_ERR_NOT_SUPPORTED;                                      \
+        }                                                                      \
+    } while(0)
+
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+ucc_status_t ucc_mc_cuda_reduce(const void *src1, const void *src2, void *dst,
+                                size_t count, ucc_datatype_t dt,
+                                ucc_reduction_op_t op)
+{
+    int th;
+    unsigned long bk;
+    cudaStream_t stream;
+
+    stream = ucc_mc_cuda.stream;
+    th = MC_CUDA_CONFIG->reduce_num_threads;
+    bk = (count + th - 1)/th;
+    if (MC_CUDA_CONFIG->reduce_num_blocks != UCC_ULUNITS_AUTO) {
+        bk = ucc_min(bk, MC_CUDA_CONFIG->reduce_num_blocks);
+    }
+    switch (dt)
+    {
+        case UCC_DT_INT16:
+            DT_REDUCE_INT(int16_t, op, src1, src2, dst, count, stream, th, bk);
+            break;
+        case UCC_DT_INT32:
+            DT_REDUCE_INT(int32_t, op, src1, src2, dst, count, stream, th, bk);
+            break;
+        case UCC_DT_INT64:
+            DT_REDUCE_INT(int64_t, op, src1, src2, dst, count, stream, th, bk);
+            break;
+        case UCC_DT_FLOAT32:
+            ucc_assert(4 == sizeof(float));
+            DT_REDUCE_FLOAT(float, op, src1, src2, dst, count, stream, th, bk);
+            break;
+        case UCC_DT_FLOAT64:
+            ucc_assert(8 == sizeof(double));
+            DT_REDUCE_FLOAT(double, op, src1, src2, dst, count, stream, th, bk);
+            break;
+        default:
+            mc_error(&ucc_mc_cuda.super, "unsupported reduction type (%d)", dt);
+            return UCC_ERR_NOT_SUPPORTED;
+    }
+    CUDACHECK(cudaStreamSynchronize(stream));
```

Comment:
maybe we need a split interface: reduce_nb + sync? can we try implementing overlap if we have such a nb reduce ?

Reply:
i think it can be useful. What reduce_nb should return? Some kind of request probably. What will return cpu reduce?